src/ir/auth_logic/universe_relation_insertion.h

```diff
@@ -0,0 +1,449 @@
+        // So to do this, we iterate through the list of arguments in the
+        // predicate and the list of parameters in the declaration using
+        // the indices of the arguments iterator of the predicate.
+        for (unsigned i = 0; i < pred.args().size(); i++) {
+          AddTyping(pred.args()[i], decl.arguments()[i].argument_type());
+        }
+        return Unit();
+      }
+    }
+
+    TypeEnv type_env_;
+    DeclEnv decl_env_;
+  };
+
+  class RelationDeclarationEnvironmentVisitor
+      : public AuthLogicAstTraversingVisitor<
+            RelationDeclarationEnvironmentVisitor> {
+   public:
+    // A map from predicate names to predicate declarations.
+    using DeclEnv =
+        absl::flat_hash_map<std::string, datalog::RelationDeclaration>;
+
+    const DeclEnv& decl_env() { return decl_env_; }
+
+   private:
+    Unit Visit(const datalog::RelationDeclaration& rel_decl) {
+      decl_env_.insert({std::string{rel_decl.relation_name()}, rel_decl});
+      return Unit();
+    }
+
+    DeclEnv decl_env_;
+  };
+
+  // This method when given a program returns a set (so the elements
+  // are unique) of relation declarations for the universes that are needed
+  // based on the types of elements that are mentioned in the original
+  // declarations of the program.
+  static absl::flat_hash_set<datalog::RelationDeclaration>
+  GetUniverseDeclarations(const Program& prog) {
+    UniverseDeclarationGenerationVisitor gen;
+    absl::flat_hash_set<datalog::RelationDeclaration> ret = prog.Accept(gen);
+    // It is possible that universes need to be created
+    // for numbers and principals even if they are not
+    // referenced in relation declarations in the front-end
+    absl::flat_hash_set<datalog::RelationDeclaration> builtin_universes = {
+        TypeToUniverseDeclaration(datalog::ArgumentType(
+            datalog::ArgumentType::Kind::kNumber, "Number")),
+        TypeToUniverseDeclaration(datalog::ArgumentType(
+            datalog::ArgumentType::Kind::kPrincipal, "Principal")),
+    };
+    ret.merge(builtin_universes);
+    return ret;
+  }
+
+  // This visitor is used only to implmenet GetUniverseDeclarations
+  class UniverseDeclarationGenerationVisitor
+      : public AuthLogicAstTraversingVisitor<
+            UniverseDeclarationGenerationVisitor,
+            absl::flat_hash_set<datalog::RelationDeclaration>> {
+   private:
+    using DeclSet = absl::flat_hash_set<datalog::RelationDeclaration>;
+    DeclSet GetDefaultValue() override { return {}; }
+
+    DeclSet CombineResult(DeclSet left, DeclSet right) override {
+      left.merge(std::move(right));
+      return left;
+    }
+
+    DeclSet Visit(
+        const datalog::RelationDeclaration& relation_declaration) override {
+      DeclSet universe_declarations;
+      for (const auto& argument : relation_declaration.arguments()) {
+        universe_declarations.emplace(
+            TypeToUniverseDeclaration(argument.argument_type()));
```

Comment:
In general, just use `insert`. You should only use `emplace` rather than `insert` when forced to by the compiler.

Reply:
fixed